File: vol131/13157.py
# TLE
N = 100000
prime, layout = [ False ] * 2 + [ True ] + [ True, False ] * ( N - 3 ), [ [ 0 ], [ 0, 2 ], [ 0, 2, 5, 7 ] ]

for i in range( 3, int( N ** .5 ), 2 ):
	if prime[ i ]:
		for j in range( i ** 2, N, 2 * i ):
			prime[ j ] = False

for i in range( 3, N, 2 ):
	if prime[ i ]:
		layout += [ [ 0, i ] ]
		if prime[ i - 2 ]:
			layout += [ [ 0, 2, i ], [ 0, i - 2, i ] ]

for t in range( int( input() ) ):
	cell, coin = map( int, input().split() )
	ans, assign = 0, [ 1 ]
	for i in range( 1, 4 ):
		assign += [ assign[ -1 ] * ( coin - i ) // i ]
	for l in layout:
		ans += assign[ len( l ) - 1 ] * max( 0, cell - l[ -1 ] )
	print( 'Case %d: %d' % ( t + 1, ans ) )

File: vol131/13157.cpp
#include<bits/stdc++.h>
using namespace std;

const int N = 100000, MOD = 1000000007;
int init[] = { 0, 2, 5, 7 }, t, coin, cell;
bool prime[ N ];
vector<vector<int>> layout;

int main()
{
	// Find primes
	fill( prime + 2, prime + N, true );
	for( int i = 4; i < N; i += 2 )
		prime[ i ] = false;
	for( int i = 3; i * i <= N; i += 2 )
		if( prime[ i ] )
			for( int j = i * i; j < N; j += i + i )
				prime[ j ] = false;

	// Find possible patterns
	layout.push_back( vector<int>( init, init + 1 ) );
	layout.push_back( vector<int>( init, init + 2 ) );
	layout.push_back( vector<int>( init, init + 4 ) );
	for( int i = 3; i < N; i += 2 )
		if( prime[ i ] )
		{
			init[ 1 ] = i;
			layout.push_back( vector<int>( init, init + 2 ) );
			if( prime[ i - 2 ] )
			{
				init[ 1 ] = 2;
				init[ 2 ] = i;
				layout.push_back( vector<int>( init, init + 3 ) );
				init[ 1 ] = i - 2;
				layout.push_back( vector<int>( init, init + 3 ) );
			}
		}

	cin >> t;
	for( int n = 1; n <= t && cin >> cell >> coin; ++n )
	{
		long long ans = 0, assign[ 4 ] = { 1 };
		for( int i = 1; i < 4; ++i )
			assign[ i ] = assign[ i - 1 ] * ( coin - i ) / i;
		for( auto l: layout )
			ans += assign[ l.size() - 1 ] % MOD * max( 0, cell - l.back() ) % MOD;
		printf( "Case %d: %lld\n", n, ans % MOD );
	}
}
